### PR TITLE
 Framework: Update code to work with Babel 7

### PR DESCRIPTION
**Pull request by gziolo**
Moved from https://github.com/WordPress/packages/pull/136.

This PR follows the migration docs for Babel 7 migration: https://new.babeljs.io/docs/en/next/v7-migration.html.

It is highly influenced by Calypso migration done a few weeks back on a much larger codebase with the help from Babel maintainers: https://github.com/Automattic/wp-calypso/pull/23424.

The most useful hint:
> - Updated `babel-core` to the [bridge](https://github.com/babel/babel-bridge) which is necessary for anything having babel as a peer dependency -- aka jest.

### Changes

1. `jest` and `jest-enzyme` were upgraded to the latest version to make them work with Babel 7.
1. This PR fixes the following Babel 6 issue we discovered when publishing packages to npm:
  > ```js
  > export * from './components';
  > export * from './api';
  > ```
  > 
  > transpiles to commonjs with the unexpected:
  > ```js
  > import _Object$defineProperty from 'babel-runtime/core-js/object/define-property';
  > import _Object$keys from 'babel-runtime/core-js/object/keys';
  > ```
  See related comment from @nerrad: 
  https://github.com/WordPress/gutenberg/issues/3955#issuecomment-394724086. 

### Testing
* `npm run dev`
* `npm run buil`
* `npm test`
* `npm run test-e2e`

All those commends should work, there should be no regressions when interacting with Gutenberg.

### CODE REVIEW

**Comment by gziolo** [comment]
Locally, one unit tests is failing:

> packages/core-data/src/test/selectors.js
>   ● Test suite failed to run
> 
>     TypeError: Cannot read property 'loadAndPersist' of undefined
> 
>       29 |     return _persist.loadAndPersist;
>       30 |   }
>     > 31 | });
>          |              ^
>       32 | Object.defineProperty(exports, "withRehydration", {
>       33 |   enumerable: true,
>       34 |   get: function get() {
> 
>       at Object.get [as loadAndPersist] (packages/data/build/index.js:31:21)
>       at packages/core-data/src/test/selectors.js:23:452
>           at Array.forEach (<anonymous>)
>       at _objectSpread (packages/core-data/src/test/selectors.js:23:392)
>       at Object.user:/Users/gziolo/PhpstormProjects/gutenberg/packages/data/build/index.js: (packages/core-data/src/test/selectors.js:4:10)
>       at Object.<anonymous> (packages/data/build/store/index.js:12:9)
>       at Object.<anonymous> (packages/data/build/index.js:84:37)
>       at Object.user:/Users/gziolo/PhpstormProjects/gutenberg/packages/data/build/index.js: (packages/core-data/src/test/selectors.js:4:36)
>       at Object.<anonymous> (packages/core-data/src/selectors.js:29:13)
>       at Object.<anonymous> (packages/core-data/src/test/selectors.js:17:18)

I will further investigate tomorrow.

File: package.json [hunk 1]
```diff
@@ -178,7 +150,7 @@
 		"build:packages": "cross-env EXCLUDE_PACKAGES=babel-plugin-import-jsx-pragma node ./bin/packages/build.js",
 		"build": "npm run build:packages && cross-env NODE_ENV=production webpack",
 		"check-engines": "check-node-version --package",
-		"ci": "concurrently \"npm run lint && npm run build\" \"npm run test-unit:coverage-ci\"",
+		"ci": "concurrently \"npm run lint\" \"npm run test-unit:coverage-ci\"",
```
Comment (gziolo) [inline package.json]: `npm run build` may accidentally regenerate `jest-console` build files which are used in framework setup script for each test suite. We run it on the node which runs e2e tests anyways.

File: packages/data/src/index.js [hunk 2]
```diff
@@ -20,7 +20,11 @@ import isShallowEqual from '@wordpress/is-shallow-equal';
  */
 import registerDataStore from './store';
 
-export { loadAndPersist, withRehydration } from './persist';
+export {
+	loadAndPersist,
+	setPersistenceStorage,
```
Comment (gziolo) [inline packages/data/src/index.js]: `setPersistenceStorage` - it probably should be another PR. I added it because it wasn't exposed.

File: packages/jest-preset-default/jest-preset.json [hunk 3]
```diff
@@ -22,8 +22,7 @@
 	],
 	"timers": "fake",
 	"transform": {
-		"^.+\\.jsx?$": "babel-jest",
-		"\\.pegjs$": "<rootDir>/node_modules/@wordpress/jest-preset-default/scripts/pegjs-transform.js"
```
Comment (gziolo) [inline packages/jest-preset-default/jest-preset.json]: `pegjs` is not needed outside of Gutenberg.

File: packages/jest-preset-default/scripts/pegjs-transform.js [hunk 4]
```diff
@@ -1,16 +0,0 @@
-const pegjs = require( 'pegjs' );
```
Comment (gziolo) [inline packages/jest-preset-default/scripts/pegjs-transform.js]: The same applies - `pegjs` transform is duplicated here and we need it only for Gutenberg.

File: packages/jest-preset-default/scripts/setup-globals.js [hunk 5]
```diff
@@ -13,35 +13,6 @@ global.window.matchMedia = () => ( {
 	removeListener: () => {},
 } );
 
-global.window._wpDateSettings = {
```
Comment (gziolo) [inline packages/jest-preset-default/scripts/setup-globals.js]: We are now providing those defaults inside `@wordpress/date` package.

File: packages/jest-preset-default/scripts/setup-globals.js [hunk 6]
```diff
@@ -51,6 +22,3 @@ global.window.localStorage = {
 
 // UserSettings global
 global.window.userSettings = { uid: 1 };
-
-// Mock jQuery
-global.window.jQuery = { holdReady() {} };
```
Comment (gziolo) [inline packages/jest-preset-default/scripts/setup-globals.js]: We no longer use `holdReady` in the codebase, besides it isn't general enough to be here.

File: test/unit/jest.config.json [hunk 7]
```diff
@@ -4,13 +4,18 @@
...
+		"<rootDir>/packages/.*/benchmark/"
+	],
 	"moduleNameMapper": {
 		"@wordpress\\/(blocks|components|editor|utils|edit-post|viewport|core-blocks|nux)$": "$1",
-		"@wordpress\\/(.*)$": "packages/$1/src"
+		"@wordpress\\/([a-z0-9-])$": "packages/$1/src"
 	},
 	"preset": "@wordpress/jest-preset-default",
 	"setupFiles": [
-		"core-js/fn/symbol/async-iterator",
```
Comment (gziolo) [inline test/unit/jest.config.json]: It's included in Babel's preset.

**Comment by gziolo** [comment]
Travis is happy about the current shape of this PR, I am as well. It should be ready to be landed once it is tested and all review feedback gets addressed.

**Comment by gziolo** [comment]
Chunks stats for `npm run build` without gzip:

### After

                                  Asset       Size  Chunks                    Chunk Names
              ./build/viewport/index.js   24.2 KiB      19  [emitted]         viewport
             ./build/shortcode/index.js   16.1 KiB       0  [emitted]         shortcode
              ./build/keycodes/index.js   17.7 KiB       2  [emitted]         keycodes
      ./build/is-shallow-equal/index.js   6.39 KiB       3  [emitted]         isShallowEqual
                  ./build/i18n/index.js   27.4 KiB       4  [emitted]         i18n
                 ./build/hooks/index.js   19.2 KiB       5  [emitted]         hooks
               ./build/element/index.js     42 KiB       6  [emitted]         element
             ./build/dom-ready/index.js  741 bytes       7  [emitted]         domReady
                   ./build/dom/index.js   22.8 KiB       8  [emitted]         dom
            ./build/deprecated/index.js   4.89 KiB       9  [emitted]         deprecated
                  ./build/date/index.js    192 KiB      10  [emitted]         date
                  ./build/data/index.js   60.6 KiB      11  [emitted]         data
             ./build/core-data/index.js   54.5 KiB      12  [emitted]         coreData
                  ./build/blob/index.js  909 bytes      13  [emitted]         blob
           ./build/api-request/index.js     21 KiB      14  [emitted]         apiRequest
                  ./build/a11y/index.js   5.15 KiB      15  [emitted]         a11y
                   ./build/nux/index.js   23.8 KiB      16  [emitted]         nux
           ./build/core-blocks/index.js    185 KiB      17  [emitted]         coreBlocks
             ./build/edit-post/index.js   91.7 KiB      18  [emitted]         editPost
               ./build/plugins/index.js   24.5 KiB       1  [emitted]         plugins
                 ./build/utils/index.js   40.9 KiB      20  [emitted]         utils
                ./build/editor/index.js    352 KiB      21  [emitted]  [big]  editor
            ./build/components/index.js    502 KiB      22  [emitted]  [big]  components
                ./build/blocks/index.js    302 KiB      23  [emitted]  [big]  blocks

**Review by hypest — APPROVED** [review]
👋 @gziolo , gave the PR (commit 1de19d609, and after merging from GB master) a try against this [RN app branch (](https://github.com/wordpress-mobile/gutenberg-mobile/pull/68) and it seems that the RN app builds and runs just fine. In other words, upgrading GB to Babel 7 doesn't seem to affect the current state of the RN app in a negative way. So, fwiw, 👍 from me!

**Comment by Shelob9** [comment]
@gziolo This is : 🔥

Is there any way to test the packages before the PR is merged? I was thinking build Gutenberg locally and then use npm link, but not sure if that works with this setup.

**Comment by gziolo** [comment]
> Is there any way to test the packages before the PR is merged? I was thinking build Gutenberg locally and then use npm link, but not sure if that works with this setup.

It might be tricky because we started using local packages `file:`. We can always fix everything after next alpha is published to npm. I still don't know what it the proper way to inform consumers of packages that they should have `@babel/runtime` installed locally. Should we put this in the readme file or set as an explicit dependency in each package?

File: codecov.yml [hunk 8]
```diff
@@ -7,7 +7,3 @@ coverage:
     patch: off
 
 comment: false
-
-ignore:
```
Comment (aduth) [inline codecov.yml]: Why is this being removed here?

File: package.json (hunk 1 above)
Comment (aduth) [reply package.json]: I'm having a hard time tracking down why we had `npm run build` here in the first place. If I were to guess, seems as though it were to test to ensure that we'd have no errors during a build. Could still be useful, though I'm not entirely sure it's critical (at least weighted against the time cost / complications).

File: packages/autop/CHANGELOG.md [hunk 9]
```diff
@@ -1,3 +1,8 @@
+## Unreleased (1.1.0)
```
Comment (aduth) [inline packages/autop/CHANGELOG.md]: I think formatting would be `## 1.1.0 (Unreleased)` for consistency with other releases?

File: packages/babel-plugin-import-jsx-pragma/test/index.js [hunk 10]
```diff
@@ -10,7 +10,7 @@ import plugin from '../src';
 
 describe( 'babel-plugin-import-jsx-pragma', () => {
 	function getTransformedCode( source, options = {} ) {
-		const { code } = transform( source, {
+		const { code } = transformSync( source, {
 			plugins: [
 				[ plugin, options ],
 				'syntax-jsx',
```
Comment (aduth) [inline packages/babel-plugin-import-jsx-pragma/test/index.js]: I think this needs to be updated to `@babel/syntax-jsx` ? I'm seeing some related failing tests.

File: packages/core-data/src/test/selectors.js [hunk 11]
```diff
@@ -9,10 +9,13 @@ import deepFreeze from 'deep-freeze';
 import { getTerms, isRequestingCategories, getEntityRecord, getEntityRecords } from '../selectors';
 import { select } from '@wordpress/data';
 
-jest.mock( '@wordpress/data', () => ( {
-	...require.requireActual( '@wordpress/data' ),
-	select: jest.fn().mockReturnValue( {} ),
-} ) );
+jest.mock( '@wordpress/data', () => {
+	return {
+		select: jest.fn().mockReturnValue( {
```
Comment (aduth) [inline packages/core-data/src/test/selectors.js]: Did we mean to drop the `...require.requireActual` ?

File: codecov.yml (hunk 8 above)
Comment (aduth) [reply codecov.yml]: Sorry, realized later it's now reflected in `jest.config.json`

**Comment by aduth** [comment]
Observing these errors trying to run tests:

```
Configuration error:
    
    Could not locate module /Users/andrew/Documents/Code/vvv/www/editor/htdocs/wp-content/plugins/gutenberg/node_modules/@wordpress/babel-plugin-import-jsx-pragma/build/index.js mapped as:
    packages/babel-plugin-import-jsx-pragma/build/index.js/src.
    
    Please check your configuration for these entries:
    {
      "moduleNameMapper": {
        "/@wordpress\/(.*)$/": "packages/$1/src"
      },
      "resolver": null
    }
```

File: packages/babel-preset-default/index.js [hunk 12]
```diff
@@ -1,22 +1,26 @@
...
+					browsers: [ 'extends @wordpress/browserslist-config' ],
+				},
+				useBuiltIns: 'usage',
+			} ],
+			isTestEnv && [ '@babel/preset-env', {
+				useBuiltIns: 'usage',
+			} ],
+		].filter( Boolean ),
+		plugins: [
+			'@babel/plugin-proposal-object-rest-spread',
```
Comment (aduth) [inline packages/babel-preset-default/index.js]: I _think_ these can be shortened to e.g. `@babel/proposal-object-rest-spread`. Maybe less explicit. Haven't confirmed.

**Comment by aduth** [comment]
Got closer to passing tests with this diff:

```diff
diff --git a/packages/babel-plugin-import-jsx-pragma/test/index.js b/packages/babel-plugin-import-jsx-pragma/test/index.js
index ede36f878..355a698f7 100644
--- a/packages/babel-plugin-import-jsx-pragma/test/index.js
+++ b/packages/babel-plugin-import-jsx-pragma/test/index.js
@@ -13,7 +13,7 @@ describe( 'babel-plugin-import-jsx-pragma', () => {
 		const { code } = transformSync( source, {
 			plugins: [
 				[ plugin, options ],
-				'syntax-jsx',
+				'@babel/syntax-jsx',
 			],
 		} );
 
diff --git a/test/unit/jest.config.json b/test/unit/jest.config.json
index 71c618443..5a4daa0fc 100644
--- a/test/unit/jest.config.json
+++ b/test/unit/jest.config.json
@@ -12,7 +12,7 @@
 	],
 	"moduleNameMapper": {
 		"@wordpress\\/(blocks|components|editor|utils|edit-post|viewport|core-blocks|nux)$": "$1",
-		"@wordpress\\/(.*)$": "packages/$1/src"
+		"@wordpress\\/(.*)$": "packages/$1"
 	},
 	"preset": "@wordpress/jest-preset-default",
 	"setupFiles": [
```

But still seeing failures as though it's doing much more transforms than expected in `import-jsx-pragma` tests:

```
    Expected: "import { createElement } from \"@wordpress/element\";
    let foo = <bar />;"
    Received: "\"use strict\";
    
    var _element = require(\"@wordpress/element\");
    
    var foo = (0, _element.createElement)(\"bar\", null);"
```

Also:

- Guessing we don't want to target the built files in the test mapping (for debugging?)
- Are we inheriting Babel configuration now in these tests? This was meant to be isolated . Tried some options to [disable `.babelrc`](https://babeljs.io/docs/en/next/babel-core#options) from the transform with no luck.

File: packages/autop/CHANGELOG.md (hunk 9 above)
Comment (gziolo) [reply packages/autop/CHANGELOG.md]: I can update all of them 👍

**Comment by gziolo** [comment]
@aduth - I had all those bugs fixed, but overrode by accident commit when trying to merge `package-lock.json` with `node-sass` upgrade ... 😱 

I recreated in https://github.com/WordPress/gutenberg/pull/7832/commits/20c8db64604e7014f65de9c5e56f7b94bc9bb1e1

File: packages/babel-preset-default/index.js (hunk 12 above)
Comment (gziolo) [reply packages/babel-preset-default/index.js]: Yes, I think they now discourage it though 🤷‍♂️

File: packages/babel-plugin-import-jsx-pragma/test/index.js (hunk 10 above)
Comment (gziolo) [reply packages/babel-plugin-import-jsx-pragma/test/index.js]: yes

File: packages/core-data/src/test/selectors.js (hunk 11 above)
Comment (gziolo) [reply packages/core-data/src/test/selectors.js]: I think I commented elsewhere, yes, it was failing when using `requireActual` because of some Babel + Enzyme internal things. `requireActual` isn't going through `moduleNameMapper` flow so it tried to load Babel build, but couldn't import it properly 🤷‍♂️

File: packages/babel-plugin-makepot/package.json [hunk 13]
```diff
@@ -18,15 +18,22 @@
...
+	"files": [
+		"build",
+		"build-module"
+	],
 	"main": "build/index.js",
 	"module": "build-module/index.js",
 	"dependencies": {
 		"gettext-parser": "^1.3.1",
-		"lodash": "4.17.10"
+		"lodash": "^4.17.10"
```
Comment (aduth) [inline packages/babel-plugin-makepot/package.json]: May have been mentioned already, but noting we're adding loose ranges for dependencies, which while not strictly within the scope of the effort is sensible given that these are libraries.

**Review by aduth — APPROVED** [review]
Looks great 👍

File: packages/babel-plugin-makepot/package.json (hunk 13 above)
Comment (gziolo) [reply packages/babel-plugin-makepot/package.json]: Yes, I had to regenerate lock file because of the amount of changes Babel introduced after they started using namespaced packages so I thought we can also include this change for everything that we consider libraries for consistency. We were mixing ranges and exact versions before.
